junit-platform-commons/src/main/java/org/junit/platform/commons/util/ClassLoaderUtils.java

```diff
@@ -50,6 +56,43 @@ public static ClassLoader getDefaultClassLoader() {
 		return ClassLoader.getSystemClassLoader();
 	}
 
+	static URL toURL(String filePath) {
+		try {
+			return Paths.get(filePath).toUri().toURL();
+		}
+		catch (MalformedURLException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * Returns a classloader that has visibility to the current classpath, but won't delegate to the parent for classes that
+	 * match the supplied filter. If you attempt to load a class using this loader that matches the filter, then
+	 * it will be loaded directly from the classpath and a new {@link Class} object created.
+	 *
+	 * <p>This is sometimes useful for testing if you have a class in your JVM that's already visible to the current
+	 * class loader, but for testing purposes you need to load it from the context of another class loader. This allows
+	 * you to emulate environments that have multiple classloaders (eg, OSGi).
+	 *
+	 * <p>For an example, see {@link ReflectionUtilsTest#findMethodByParameterTypesFromForeignClassLoader}.
+	 *
+	 * @param filter predicate to match on the type names. If the predicate returns true, then the classloader won't
+	 * delegate to the parent classloader but will load the class directly from the classpath.
+	 * @return The excluding class loader.
+	 */
+	public static URLClassLoader excludingClassLoader(Predicate<String> filter) {
+		return new URLClassLoader(Stream.of(System.getProperty("java.class.path").split(File.pathSeparator)).map(
+			ClassLoaderUtils::toURL).toArray(URL[]::new), new ClassLoader() {
+				@Override
+				protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+					if (filter.test(name)) {
+						throw new ClassNotFoundException("Skipping " + name + " for testing");
+					}
+					return super.loadClass(name, resolve);
+				}
+			});
+	}
+
```

Comment:
Why did you make these two new methods part of the production code even though they're only used in tests?

Reply:
The Javadoc for `ClassLoaderUtils` said it was for internal use only.

I'm happy to take guidance on a better place to put it - I wasn't sure where to put test utils. That said, an excluding class loader is also a useful thing to have around when you're doing testing that involves class loaders, so longer term maybe it's not such a silly idea to have it in the production code?

Reply:
We already have several custom class loaders used in tests.

- org.junit.jupiter.api.AssertThrowsAssertionsTests.EnigmaClassLoader
- org.junit.jupiter.engine.extension.TestInstanceFactoryTests.ProxyClassLoader
- org.junit.jupiter.engine.support.OpenTest4JAndJUnit4AwareThrowableCollectorTests.TestClassLoader

So it might be good to have some central implementations for various testing scenarios, but I don't think that necessarily warrants being in "production" code.

We could also make use of Gradle's "test fixtures" support.

Reply:
That is fair. Let me know where you'd like the testing utilities to go and I'll look at moving it there.